### DISCUSSION TITLE
For people with install issues on Raspberry Pi OS

### DISCUSSION THREAD

**Discussion by Swissbandit** [markqvist/Reticulum#115] [Help & Questions] [answered]
A Colleague had some issues getting everything installed;
pyOpenSSL < 22.0.0 seems to be incompatible since they did not pin the cryptography package. 
Since the latest cryptography package was installed in version prior 22.0.0 they simply break.
Corresponding Github issue: https://github.com/pyca/pyopenssl/issues/1143
My pip stopped working due to the issue (I couldn't even get the version).

I resolved this issue by download the problem package from pypi.org and update it via easy_install module:
python3 -m easy_install pyOpenSSL-22.0.0-py2.py3-none-any.whl
(this maybe not available any longer)
Or try this
Error with pyOpenSSL==19.0.0
and was able to fix this with
pip3 install pyOpenSSL --upgrade

lately i had no luck with it so i did this:
i had nomad installed already,
and with sudo i was able to install sbapp

**Comment by markqvist** [comment]
Interesting, thanks for posting those tips. What platform/OS is this happening on?

**Reply by Swissbandit** [reply]
Here`s what i got on a fresh official Raspi OS 64bit

fdg@raspberrypi:~ $ pip install nomadnet
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting nomadnet
  Downloading nomadnet-0.2.4-1.tar.gz (479 kB)
     |████████████████████████████████| 479 kB 1.5 MB/s 
WARNING: Discarding https://files.pythonhosted.org/packages/7a/19/d8f3e91960ceea2f0cc44ebe0ce3ad6493b8d1ef669016a2131bdea8d4c8/nomadnet-0.2.4-1.tar.gz#sha256=4c44822ea86a62c69d07f155e59e4747b18b2cc9d49994e64e2cac733428bc07 (from https://pypi.org/simple/nomadnet/) (requires-python:>=3.6). Requested nomadnet from https://files.pythonhosted.org/packages/7a/19/d8f3e91960ceea2f0cc44ebe0ce3ad6493b8d1ef669016a2131bdea8d4c8/nomadnet-0.2.4-1.tar.gz#sha256=4c44822ea86a62c69d07f155e59e4747b18b2cc9d49994e64e2cac733428bc07 has different version in metadata: '0.2.4'
  Downloading nomadnet-0.2.4-1-py3-none-any.whl (132 kB)
     |████████████████████████████████| 132 kB 6.4 MB/s 
Collecting rns>=0.3.14
  Downloading https://www.piwheels.org/simple/rns/rns-0.3.14-py3-none-any.whl (237 kB)
     |████████████████████████████████| 237 kB 1.1 MB/s 
Collecting lxmf>=0.2.0
  Downloading https://www.piwheels.org/simple/lxmf/lxmf-0.2.0-py3-none-any.whl (22 kB)
Collecting urwid>=2.1.2
  Downloading urwid-2.1.2.tar.gz (634 kB)
     |████████████████████████████████| 634 kB 12.9 MB/s 
Collecting netifaces
  Downloading netifaces-0.11.0.tar.gz (30 kB)
Collecting pyserial>=3.5
  Downloading https://www.piwheels.org/simple/pyserial/pyserial-3.5-py2.py3-none-any.whl (90 kB)
     |████████████████████████████████| 90 kB 1.0 MB/s 
Collecting cryptography>=3.4.7
  Downloading cryptography-38.0.2-cp36-abi3-manylinux_2_28_aarch64.whl (4.0 MB)
     |████████████████████████████████| 4.0 MB 2.5 MB/s 
Collecting cffi>=1.12
  Downloading cffi-1.15.1-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl (448 kB)
     |████████████████████████████████| 448 kB 2.5 MB/s 
Collecting pycparser
  Downloading https://www.piwheels.org/simple/pycparser/pycparser-2.21-py2.py3-none-any.whl (119 kB)
     |████████████████████████████████| 119 kB 1.0 MB/s 
Building wheels for collected packages: urwid, netifaces
  Building wheel for urwid (setup.py) ... done
  Created wheel for urwid: filename=urwid-2.1.2-cp39-cp39-linux_aarch64.whl size=255325 sha256=03bea5fd229a95858991bac4742b2d1b53e7afafc8ca3f8b5222e7175a288671
  Stored in directory: /home/fdg/.cache/pip/wheels/44/ec/04/2c1080c3ee4e80e76d662ac35f0594a2a86f9df12095b05cb3
  Building wheel for netifaces (setup.py) ... done
  Created wheel for netifaces: filename=netifaces-0.11.0-cp39-cp39-linux_aarch64.whl size=34931 sha256=b810dfcb2a2caccb2d302a75183d533f3aa21d49cfc413cb0ecae992361f8418
  Stored in directory: /home/fdg/.cache/pip/wheels/4f/77/60/d721ffbafc28e5021f94207ac09dbc23ecde96b4f74a324106
Successfully built urwid netifaces
Installing collected packages: pycparser, cffi, pyserial, netifaces, cryptography, rns, urwid, lxmf, nomadnet
  WARNING: The scripts pyserial-miniterm and pyserial-ports are installed in '/home/fdg/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The scripts rncp, rnpath, rnprobe, rnsd, rnstatus and rnx are installed in '/home/fdg/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The script nomadnet is installed in '/home/fdg/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed cffi-1.15.1 cryptography-38.0.2 lxmf-0.2.0 netifaces-0.11.0 nomadnet-0.2.4 pycparser-2.21 pyserial-3.5 rns-0.3.14 urwid-2.1.2


fdg@raspberrypi:~ $ pip install sbapp
Traceback (most recent call last):
  File "/usr/bin/pip", line 33, in <module>
    sys.exit(load_entry_point('pip==20.3.4', 'console_scripts', 'pip')())
  File "/usr/bin/pip", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/usr/lib/python3.9/importlib/metadata.py", line 77, in load
    module = import_module(match.group('module'))
  File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 790, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/usr/lib/python3/dist-packages/pip/_internal/cli/main.py", line 10, in <module>
    from pip._internal.cli.autocompletion import autocomplete
  File "/usr/lib/python3/dist-packages/pip/_internal/cli/autocompletion.py", line 9, in <module>
    from pip._internal.cli.main_parser import create_main_parser
  File "/usr/lib/python3/dist-packages/pip/_internal/cli/main_parser.py", line 7, in <module>
    from pip._internal.cli import cmdoptions
  File "/usr/lib/python3/dist-packages/pip/_internal/cli/cmdoptions.py", line 23, in <module>
    from pip._vendor.packaging.utils import canonicalize_name
  File "/usr/lib/python3/dist-packages/pip/_vendor/__init__.py", line 94, in <module>
    vendored("requests.packages.urllib3.contrib.pyopenssl")
  File "/usr/lib/python3/dist-packages/pip/_vendor/__init__.py", line 36, in vendored
    __import__(modulename, globals(), locals(), level=0)
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
  File "<frozen zipimport>", line 259, in load_module
  File "/usr/share/python-wheels/urllib3-1.26.5-py2.py3-none-any.whl/urllib3/contrib/pyopenssl.py", line 50, in <module>
  File "/usr/lib/python3/dist-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import crypto, SSL
  File "/usr/lib/python3/dist-packages/OpenSSL/crypto.py", line 1556, in <module>
    class X509StoreFlags(object):
  File "/usr/lib/python3/dist-packages/OpenSSL/crypto.py", line 1577, in X509StoreFlags
    CB_ISSUER_CHECK = _lib.X509_V_FLAG_CB_ISSUER_CHECK
AttributeError: module 'lib' has no attribute 'X509_V_FLAG_CB_ISSUER_CHECK'
fdg@raspberrypi:~ $ 

fdg@raspberrypi:~ $ sudo pip install sbapp
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting sbapp
  Downloading https://www.piwheels.org/simple/sbapp/sbapp-0.2.2-py3-none-any.whl (3.0 MB)
     |████████████████████████████████| 3.0 MB 1.1 MB/s 
Collecting kivy==2.1.0
  Downloading Kivy-2.1.0-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl (30.5 MB)
     |████████████████████████████████| 30.5 MB 39 kB/s 
Requirement already satisfied: pillow in /usr/lib/python3/dist-packages (from sbapp) (8.1.2)
Collecting plyer
  Downloading https://www.piwheels.org/simple/plyer/plyer-2.0.0-py2.py3-none-any.whl (131 kB)
     |████████████████████████████████| 131 kB 1.1 MB/s 
Collecting lxmf>=0.2.0
  Downloading https://www.piwheels.org/simple/lxmf/lxmf-0.2.0-py3-none-any.whl (22 kB)
Collecting rns>=0.3.14
  Downloading https://www.piwheels.org/simple/rns/rns-0.3.14-py3-none-any.whl (237 kB)
     |████████████████████████████████| 237 kB 1.0 MB/s 
Requirement already satisfied: pygments in /usr/lib/python3/dist-packages (from kivy==2.1.0->sbapp) (2.7.1)
Requirement already satisfied: docutils in /usr/lib/python3/dist-packages (from kivy==2.1.0->sbapp) (0.16)
Collecting Kivy-Garden>=0.1.4
  Downloading Kivy_Garden-0.1.5-py3-none-any.whl (4.6 kB)
Requirement already satisfied: requests in /usr/lib/python3/dist-packages (from Kivy-Garden>=0.1.4->kivy==2.1.0->sbapp) (2.25.1)
Collecting pyserial>=3.5
  Downloading https://www.piwheels.org/simple/pyserial/pyserial-3.5-py2.py3-none-any.whl (90 kB)
     |████████████████████████████████| 90 kB 1.8 MB/s 
Collecting netifaces
  Downloading netifaces-0.11.0.tar.gz (30 kB)
Collecting cryptography>=3.4.7
  Downloading cryptography-38.0.2-cp36-abi3-manylinux_2_28_aarch64.whl (4.0 MB)
     |████████████████████████████████| 4.0 MB 2.1 MB/s 
Collecting cffi>=1.12
  Downloading cffi-1.15.1-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl (448 kB)
     |████████████████████████████████| 448 kB 17.1 MB/s 
Collecting pycparser
  Downloading https://www.piwheels.org/simple/pycparser/pycparser-2.21-py2.py3-none-any.whl (119 kB)
     |████████████████████████████████| 119 kB 1.0 MB/s 
Building wheels for collected packages: netifaces
  Building wheel for netifaces (setup.py) ... done
  Created wheel for netifaces: filename=netifaces-0.11.0-cp39-cp39-linux_aarch64.whl size=34926 sha256=0ac6e1b7285abd7e17a75417e9575eb4d200723dcec62371baafb7d5161c0a39
  Stored in directory: /root/.cache/pip/wheels/4f/77/60/d721ffbafc28e5021f94207ac09dbc23ecde96b4f74a324106
Successfully built netifaces
Installing collected packages: pycparser, cffi, pyserial, netifaces, cryptography, rns, Kivy-Garden, plyer, lxmf, kivy, sbapp
  Attempting uninstall: pyserial
    Found existing installation: pyserial 3.5b0
    Not uninstalling pyserial at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'pyserial'. No files were found to uninstall.
  Attempting uninstall: cryptography
    Found existing installation: cryptography 3.3.2
    Not uninstalling cryptography at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'cryptography'. No files were found to uninstall.
Successfully installed Kivy-Garden-0.1.5 cffi-1.15.1 cryptography-38.0.2 kivy-2.1.0 lxmf-0.2.0 netifaces-0.11.0 plyer-2.0.0 pycparser-2.21 pyserial-3.5 rns-0.3.14 sbapp-0.2.2

fdg@raspberrypi:~ $ sideband
[WARNING] [Config      ] Older configuration version detected (0 instead of 24)
[WARNING] [Config      ] Upgrading configuration in progress.
[INFO   ] [Logger      ] Record log in /home/fdg/.kivy/logs/kivy_22-10-12_0.txt
[INFO   ] [Kivy        ] v2.1.0
[INFO   ] [Kivy        ] Installed at "/usr/local/lib/python3.9/dist-packages/kivy/__init__.py"
[INFO   ] [Python      ] v3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110]
[INFO   ] [Python      ] Interpreter at "/usr/bin/python3"
[INFO   ] [Logger      ] Purge log fired. Processing...
[INFO   ] [Logger      ] Purge finished!
[2022-10-12 02:49:43] [Notice] Creating new Sideband configuration...
[2022-10-12 02:49:48] [Notice] Launching platform-specific service for RNS and LXMF
[2022-10-12 02:49:48] [Notice] Purging stale messages... /home/fdg/.config/sideband/app_storage/sideband.db
fdg@raspberrypi:~ $

**Reply by markqvist** [reply]
Thanks for sharing that! I was able to get to the bottom of it, and posted a solution below.

**Comment by markqvist** [comment, marked as answer]
## Problem
Ok, please do **not** to mess around with openSSL packages through `easy_install`. It is likely to break stuff on your system later ;)

This issue is occurring because even the latest Raspberry Pi OS ships a *very* old version of `pip`, that makes calls to a *very* old version of the `cryptography` API. Raspberry Pi OS should update their shipped `pip` version.

When Reticulum or related programs are installed, the `cryptography` package is updated to a newer version, and the ancient `pip` breaks.

## Solution
The **correct (and easy) solution** here is simply to update `pip` before installing any software that requires a semi-recent version of the `cryptography` package:

```
# Update pip
pip install pip --upgrade

# Then install Reticulum
pip install rns

# Or install nomadnet
pip install nomadnet
```

Also, remember that if this is the *very first time* you install something with pip, you will need to reboot the system (or at least log out of, and back into your session) before the commands become available in the terminal (the PATH to locally installed programs is not set up before the first installation).

TL;DR: Update `pip` to a recent version before installing, and all is well.

**Reply by Swissbandit** [reply]
Thanks a lot,its usually the simplest things that get overlooked.
